### PR TITLE
Adding PredictHelper class and unit tests

### PR DESCRIPTION
**Pull request by freddyaboulton**
Adds a PredictHelper class for getting the future and past of an agent given its instance token and one of its sample tokens. 

@holger-nutonomy, I've created a base branch for the challenge called `nuscenes-predict-challenge`. Let me know if this is ok with you.

How to test:
`nosetests nuscenes/tests`

FYI: @emwolff

### CODE REVIEW

File: python-sdk/nuscenes/predict/helper.py [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+from pyquaternion import Quaternion
+from nuscenes import NuScenes
+from nuscenes.utils.geometry_utils import transform_matrix
+from nuscenes.eval.common.utils import quaternion_yaw
+import numpy as np
+
+MICROSECONDS_PER_SECOND = 1e6
+BUFFER = 0.15 # seconds
+
+def angle_of_rotation(yaw: float) -> float:
```
Comment (emwolff) [inline python-sdk/nuscenes/predict/helper.py]: @freddyaboulton Maybe a note on the conventions we are converting between?

File: python-sdk/nuscenes/predict/helper.py [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+    return (np.pi / 2) + np.sign(-yaw) * np.abs(yaw)
+
+def convert_global_coords_to_local(coordinates: np.ndarray,
+                                   translation: Tuple[float, float, float],
+                                   rotation: Tuple[float, float, float, float]) -> np.ndarray:
+    """Converts global coordinates to coordinates in the frame given by the rotation quaternion and
+    centered at the translation vector.
+    :param coordinates: x,y locations. array of shape [n_steps, 2]
+    :param translation: Tuple of (x, y, z) location that is the center of the new frame
+    :param rotation: Tuple representation of quaternion of new frame."""
```
Comment (emwolff) [inline python-sdk/nuscenes/predict/helper.py]: @freddyaboulton Do we have a clearly defined representation for a quaternion? There are two major conventions.

Comment (emwolff) [reply python-sdk/nuscenes/predict/helper.py]: Also, this is just for rotations around the z axis, no? Should be clear that it is not a general rotation routine.

File: python-sdk/nuscenes/predict/helper.py [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+    def _map_sample_and_instance_to_annotation(self) -> Dict[Tuple[str, str], str]:
+        """Creates mapping to look up an annotation given a sample and instance in constant time."""
+        mapping = {}
+
+        for record in self.data.sample_annotation:
+            mapping[(record['sample_token'], record['instance_token'])] = record['token']
+
+        return mapping
+
+    def _timestamp_for_sample(self, sample: str):
```
Comment (emwolff) [inline python-sdk/nuscenes/predict/helper.py]: Type hint for return missing

File: python-sdk/nuscenes/predict/helper.py [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+        return self.data.get('sample', sample)['timestamp']
+
+    def _absolute_time_diff(self, time1: int, time2: int) -> int:
+        """Helper to compute how much time has elapsed in _iterate method."""
+        return abs(time1 - time2) / MICROSECONDS_PER_SECOND
+
+    def _iterate(self, starting_annotation: Dict[str, Any], seconds: float, direction: str) -> List[Dict[str, Any]]:
+        """Iterates forwards or backwards in time through the annotations for a given amount of seconds."""
+
+        seconds_with_buffer = seconds + BUFFER
```
Comment (emwolff) [inline python-sdk/nuscenes/predict/helper.py]: Might want to do a sanity check that `seconds` is non-negative.

File: python-sdk/nuscenes/predict/helper.py [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+        return annotations
+
+    def get_sample_annotation(self, instance: str, sample: str) -> Dict[str, Any]:
+        """Retrives an annotation given an instance token and its sample.
+        :param instance: Instance token
+        :param sample: Sample token
+        """
+        return self.data.get('sample_annotation', self.inst_sample_to_ann[(sample, instance)])
+
+    def get_future_for_agent(self, instance: str, sample: str,
```
Comment (emwolff) [inline python-sdk/nuscenes/predict/helper.py]: @freddyaboulton You have a ton of duplicated code between this and `get_past_for_agent`. Can you introduce a function that is called by each but with the appropriate `next` or `prev` flag?

Comment (emwolff) [reply python-sdk/nuscenes/predict/helper.py]: This function would be called internally by `get_future_for_agent` and  `get_past_for_agent`.

File: python-sdk/nuscenes/predict/helper.py [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+        coords = np.array([r['translation'][:2] for r in past])
+
+        if in_agent_frame:
+            coords = convert_global_coords_to_local(coords,
+                                                    starting_annotation['translation'],
+                                                    starting_annotation['rotation'])
+
+        return coords
+
+    def get_future_for_sample(self, sample: str, seconds: float, in_agent_frame: bool) -> Dict[str, np.ndarray]:
```
Comment (emwolff) [inline python-sdk/nuscenes/predict/helper.py]: Likewise.

File: python-sdk/nuscenes/predict/helper.py [hunk 7]
```diff
@@ -0,0 +1,181 @@
+
```
Comment (emwolff) [inline python-sdk/nuscenes/predict/helper.py]: Should have our copyright blurb.

File: python-sdk/nuscenes/tests/test_predict_helper.py [hunk 8]
```diff
@@ -0,0 +1,475 @@
+import unittest
```
Comment (emwolff) [inline python-sdk/nuscenes/tests/test_predict_helper.py]: Missing copyright blurb.

File: python-sdk/nuscenes/tests/test_predict_helper.py [hunk 9]
```diff
@@ -0,0 +1,475 @@
...
+
+    @property
+    def sample_annotation(self,) -> List[Dict[str, Any]]:
+        return list(self._sample_annotation.values())
+
+    def get(self, table_name: str, token: str) -> Dict[str, Any]:
+        assert table_name in {'sample_annotation', 'sample'}
+        return getattr(self, "_" + table_name)[token]
+
+class Test_convert_global_coords_to_local(unittest.TestCase):
```
Comment (emwolff) [inline python-sdk/nuscenes/tests/test_predict_helper.py]: Excellent tests!

File: python-sdk/nuscenes/predict/helper.py (hunk 1 above)
Comment (freddyaboulton) [reply python-sdk/nuscenes/predict/helper.py]: Done

File: python-sdk/nuscenes/predict/helper.py (hunk 2 above)
Comment (freddyaboulton) [reply python-sdk/nuscenes/predict/helper.py]: Added a bit of info on the quaternion representation but it is the same as the one stored in the nuScenes dataset.

File: python-sdk/nuscenes/predict/helper.py (hunk 3 above)
Comment (freddyaboulton) [reply python-sdk/nuscenes/predict/helper.py]: Done

File: python-sdk/nuscenes/predict/helper.py (hunk 4 above)
Comment (freddyaboulton) [reply python-sdk/nuscenes/predict/helper.py]: Done!

File: python-sdk/nuscenes/predict/helper.py (hunk 5 above)
Comment (freddyaboulton) [reply python-sdk/nuscenes/predict/helper.py]: Done!

File: python-sdk/nuscenes/tests/test_predict_helper.py (hunk 8 above)
Comment (freddyaboulton) [reply python-sdk/nuscenes/tests/test_predict_helper.py]: Added the copyright blurb like the other nuscenes files.

File: python-sdk/nuscenes/predict/helper.py (hunk 7 above)
Comment (freddyaboulton) [reply python-sdk/nuscenes/predict/helper.py]: Done

File: python-sdk/nuscenes/tests/test_predict_helper.py [hunk 10]
```diff
@@ -0,0 +1,503 @@
+# nuScenes dev-kit.
```
Comment (holger-nutonomy) [inline python-sdk/nuscenes/tests/test_predict_helper.py]: I would recommend a local test folder inside the current dir (`python-sdk/nuscenes/predict`).

File: python-sdk/nuscenes/predict/helper.py [hunk 11]
```diff
@@ -0,0 +1,195 @@
...
+        for record in self.data.sample_annotation:
+            mapping[(record['sample_token'], record['instance_token'])] = record['token']
+
+        return mapping
+
+    def _timestamp_for_sample(self, sample: str) -> float:
+        """Gets timestamp from sample token."""
+        return self.data.get('sample', sample)['timestamp']
+
+    def _absolute_time_diff(self, time1: int, time2: int) -> int:
```
Comment (holger-nutonomy) [inline python-sdk/nuscenes/predict/helper.py]: Should the return type be a float?

Comment (holger-nutonomy) [reply python-sdk/nuscenes/predict/helper.py]: Same for the inputs, PyCharm says they're floats.

File: python-sdk/nuscenes/predict/helper.py [hunk 12]
```diff
@@ -0,0 +1,195 @@
...
+
+        if in_agent_frame:
+            coords = convert_global_coords_to_local(coords,
+                                                    starting_annotation['translation'],
+                                                    starting_annotation['rotation'])
+
+        return coords
+
+    def get_future_for_agent(self, instance: str, sample: str,
+                             seconds: float, in_agent_frame: bool) -> np.array:
```
Comment (holger-nutonomy) [inline python-sdk/nuscenes/predict/helper.py]: No docstring for in_agent_frame

File: python-sdk/nuscenes/predict/helper.py [hunk 13]
```diff
@@ -0,0 +1,195 @@
...
+                                seconds, in_agent_frame)
+
+            sequences[annotation_record['instance_token']] = sequence
+
+        return sequences
+
+
+    def get_future_for_sample(self, sample: str, seconds: float, in_agent_frame: bool) -> Dict[str, np.ndarray]:
+        """Retrieves the the future x,y locations of all agents in the sample.
+        :param instance: Instance token
```
Comment (holger-nutonomy) [inline python-sdk/nuscenes/predict/helper.py]: instance no longer exists, in_agent_frame is missing the param keyword

File: python-sdk/nuscenes/predict/helper.py [hunk 14]
```diff
@@ -0,0 +1,195 @@
...
+from nuscenes.utils.geometry_utils import transform_matrix
+from nuscenes.eval.common.utils import quaternion_yaw
+import numpy as np
+
+MICROSECONDS_PER_SECOND = 1e6
+BUFFER = 0.15 # seconds
+
+def angle_of_rotation(yaw: float) -> float:
+    """Given a yaw (measured from x axis) find the angle needed to rotate by so that
+    the yaw points upwards (pi / 2)."""
```
Comment (holger-nutonomy) [inline python-sdk/nuscenes/predict/helper.py]: Yaw is in the xy plane. How can it point upwards?

File: python-sdk/nuscenes/tests/test_predict_helper.py [hunk 15]
```diff
@@ -0,0 +1,503 @@
...
+
+        nusc = MockNuScenes(self.multiagent_mock_annotations, mock_samples)
+        helper = PredictHelper(nusc)
+        past = helper.get_past_for_sample('5', 3, True)
+
+        answer = {'1': np.array([[-1, -1], [-2, -2], [-3, -3]]),
+                  '2': np.array([[-1, -1], [-2, -2], [-3, -3]])}
+
+        for k in answer:
+            np.testing.assert_equal(answer[k], answer[k])
```
Comment (holger-nutonomy) [inline python-sdk/nuscenes/tests/test_predict_helper.py]: `np.testing.assert_equal(answer[k], past[k])` ?

File: python-sdk/nuscenes/tests/test_predict_helper.py [hunk 16]
```diff
@@ -0,0 +1,503 @@
+# nuScenes dev-kit.
+# Code written by Freddy Boulton, 2020.
+
+import unittest
+from typing import Dict, List, Any
+from nuscenes.predict import PredictHelper, convert_global_coords_to_local
+import numpy as np
+
+class MockNuScenes:
```
Comment (holger-nutonomy) [inline python-sdk/nuscenes/tests/test_predict_helper.py]: Propose to use inheritance here to make it cleaner. `MockNuscenes` simply is a NuScenes subclass that overwrites some methods.

File: python-sdk/nuscenes/tests/test_predict_helper.py [hunk 17]
```diff
@@ -0,0 +1,503 @@
...
+        along_neg_y[:, 1] = -np.arange(1, 6)
+        self.along_pos_x, self.along_pos_y = along_pos_x, along_pos_y
+        self.along_neg_x, self.along_neg_y = along_neg_x, along_neg_y
+
+        y_equals_x = np.zeros((5, 2))
+        y_equals_x[:, 0] = np.arange(1, 6)
+        y_equals_x[:, 1] = np.arange(1, 6)
+        self.y_equals_x = y_equals_x
+
+    def test_heading_0(self):
```
Comment (holger-nutonomy) [inline python-sdk/nuscenes/tests/test_predict_helper.py]: These 6 test methods each have 19-33 lines of redundant code. I suggest to create a separate method.

File: python-sdk/nuscenes/tests/test_predict_helper.py [hunk 18]
```diff
@@ -0,0 +1,503 @@
...
+        return list(self._sample_annotation.values())
+
+    def get(self, table_name: str, token: str) -> Dict[str, Any]:
+        assert table_name in {'sample_annotation', 'sample'}
+        return getattr(self, "_" + table_name)[token]
+
+class Test_convert_global_coords_to_local(unittest.TestCase):
+
+    def setUp(self):
+        along_pos_x = np.zeros((5, 2))
```
Comment (holger-nutonomy) [inline python-sdk/nuscenes/tests/test_predict_helper.py]: Some comments on what these are may be helpful.

File: python-sdk/nuscenes/tests/test_predict_helper.py [hunk 19]
```diff
@@ -0,0 +1,503 @@
...
+
+        # Testing path along neg y direction
+        answer = convert_global_coords_to_local(self.along_neg_y, origin, rotation)
+        np.testing.assert_allclose(answer, self.along_pos_x, atol=1e-4)
+
+        answer = convert_global_coords_to_local(self.along_neg_y + [[50, 25]], offset, rotation)
+        np.testing.assert_allclose(answer, self.along_pos_x, atol=1e-4)
+
+    def test_heading_pi_over_4(self):
+        rotation = (np.cos(np.pi / 8), 0, 0, np.sin(np.pi / 8))
```
Comment (holger-nutonomy) [inline python-sdk/nuscenes/tests/test_predict_helper.py]: The method names and the rotations are off by factor 2. Why is that?

File: python-sdk/nuscenes/predict/helper.py [hunk 20]
```diff
@@ -0,0 +1,235 @@
...
+
+        for record in self.data.sample_annotation:
+            mapping[(record['sample_token'], record['instance_token'])] = record['token']
+
+        return mapping
+
+    def _timestamp_for_sample(self, sample: str) -> float:
+        """
+        Gets timestamp from sample token.
+        :param sample: TODO.
```
Comment (holger-nutonomy) [inline python-sdk/nuscenes/predict/helper.py]: I suggest to add `_token` to each token to make it clearer that this is not the sample dict.

File: python-sdk/nuscenes/predict/helper.py (hunk 11 above)
Comment (freddyaboulton) [reply python-sdk/nuscenes/predict/helper.py]: Thanks for the change - I thought the timestamps were kept as ints in nuscenes.

File: python-sdk/nuscenes/predict/helper.py (hunk 14 above)
Comment (freddyaboulton) [reply python-sdk/nuscenes/predict/helper.py]: Updated the comment. Upwards = aligned with y-xais :sweat_smile:

File: python-sdk/nuscenes/tests/test_predict_helper.py (hunk 10 above)
Comment (freddyaboulton) [reply python-sdk/nuscenes/tests/test_predict_helper.py]: Done!

File: python-sdk/nuscenes/tests/test_predict_helper.py (hunk 19 above)
Comment (freddyaboulton) [reply python-sdk/nuscenes/tests/test_predict_helper.py]: I'm using this convention `cos(theta / 2) + (xi + yi + zi)sin(theta / 2)` to convert between yaw and quaternion

File: python-sdk/nuscenes/tests/test_predict_helper.py (hunk 18 above)
Comment (freddyaboulton) [reply python-sdk/nuscenes/tests/test_predict_helper.py]: Done!

File: python-sdk/nuscenes/tests/test_predict_helper.py (hunk 15 above)
Comment (freddyaboulton) [reply python-sdk/nuscenes/tests/test_predict_helper.py]: Fixed!

File: python-sdk/nuscenes/tests/test_predict_helper.py (hunk 17 above)
Comment (freddyaboulton) [reply python-sdk/nuscenes/tests/test_predict_helper.py]: Done!

**Comment by freddyaboulton** [comment]
@holger-nutonomy Addressed your comments! Thanks for the feedback.

File: python-sdk/nuscenes/tests/test_predict_helper.py (hunk 19 above)
Comment (holger-nutonomy) [reply python-sdk/nuscenes/tests/test_predict_helper.py]: We usually use
- [quaternion_yaw](https://github.com/nutonomy/nuscenes-devkit/blob/053438468776a51f4a522cbf02023d54efdeaac3/python-sdk/nuscenes/eval/common/utils.py#L112)
- `q = Quaternion(axis=(0, 0, 1), angle=yaw_in)`

**Pull request merged**